mr/src/main/java/org/apache/iceberg/mr/mapred/IcebergFilterFactory.java

```diff
@@ -0,0 +1,151 @@
+   * @param leaves List of all leaf nodes within the tree.
+   * @return Expression that is translated from the Hive SearchArgument.
+   */
+  private static Expression translate(ExpressionTree tree, List<PredicateLeaf> leaves) {
+    List<ExpressionTree> childNodes = tree.getChildren();
+    switch (tree.getOperator()) {
+      case OR:
+        Expression orResult = Expressions.alwaysFalse();
+        for (ExpressionTree child : childNodes) {
+          orResult = or(orResult, translate(child, leaves));
+        }
+        return orResult;
+      case AND:
+        Expression result = Expressions.alwaysTrue();
+        for (ExpressionTree child : childNodes) {
+          result = and(result, translate(child, leaves));
+        }
+        return result;
+      case NOT:
+        return not(translate(tree.getChildren().get(0), leaves));
+      case LEAF:
+        return translateLeaf(leaves.get(tree.getLeaf()));
+      case CONSTANT:
+        //We are unsure of how the CONSTANT case works, so using the approach of:
+        //https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/read/
+        // ParquetFilterPredicateConverter.java#L116
+        throw new UnsupportedOperationException("CONSTANT operator is not supported");
+      default:
+        throw new IllegalStateException("Unknown operator: " + tree.getOperator());
+    }
+  }
+
+  /**
+   * Translate leaf nodes from Hive operator to Iceberg operator.
+   * @param leaf Leaf node
+   * @return Expression fully translated from Hive PredicateLeaf
+   */
+  private static Expression translateLeaf(PredicateLeaf leaf) {
+    String column = leaf.getColumnName();
+    switch (leaf.getOperator()) {
+      case EQUALS:
+        return equal(column, leafToIcebergType(leaf));
+      case NULL_SAFE_EQUALS:
+        return equal(notNull(column).ref().name(), leafToIcebergType(leaf)); //TODO: Unsure..
+      case LESS_THAN:
+        return lessThan(column, leafToIcebergType(leaf));
+      case LESS_THAN_EQUALS:
+        return lessThanOrEqual(column, leafToIcebergType(leaf));
+      case IN:
+        return in(column, hiveLiteralListToIcebergType(leaf.getLiteralList()));
+      case BETWEEN:
+        List<Object> icebergLiterals = hiveLiteralListToIcebergType(leaf.getLiteralList());
+        return and(greaterThanOrEqual(column, icebergLiterals.get(0)),
+            lessThanOrEqual(column, icebergLiterals.get(1)));
+      case IS_NULL:
+        return isNull(column);
+      default:
+        throw new IllegalStateException("Unknown operator: " + leaf.getOperator());
+    }
+  }
+
+  private static Object leafToIcebergType(PredicateLeaf leaf) {
+    switch (leaf.getType()) {
+      case LONG:
+        return leaf.getLiteral();
+      case FLOAT:
+        return leaf.getLiteral();
+      case STRING:
+        return leaf.getLiteral();
+      case DATE:
+        return ((Timestamp) leaf.getLiteral()).getTime();
```

Comment:
IIRC Hive's Date type predicate literal is of type java.sql.Date. So this looks incorrect? (Unsure)

Reply:
Yeah you're correct about the Date type, and I was using that initially, but then when running the tests I would get a `ClassCastException` as it looks like Hive does some internal conversion into a Timestamp when calling leaf.getLiteral for a Date type: 
https://github.com/apache/hive/blob/branch-2.3/storage-api/src/java/org/apache/hadoop/hive/ql/io/sarg/SearchArgumentImpl.java#L108

Reply:
Ugh! Nice find. `Timestamp.getTime()` is probably still incorrect though as Iceberg expects the number of days since epoch as a literal for the Date type. `Timestamp.getTime()` will give you the number of mills since epoch.

Reply:
Also, might be good to mention this issue as a comment as it looks very non-intuitive.

Reply:
Ah yeah, good shout on the comment and using the correct granularity, I'll make those changes!